src/main/java/org/jboss/threads/CachedExecutorView.java

```diff
@@ -0,0 +1,298 @@
+/*
+ * JBoss, Home of Professional Open Source.
+ * Copyright 2020 Red Hat, Inc., and individual contributors
+ * as indicated by the @author tags.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.jboss.threads;
+
+import org.wildfly.common.Assert;
+import org.wildfly.common.annotation.Nullable;
+
+import java.util.Collections;
+import java.util.List;
+import java.util.Set;
+import java.util.concurrent.AbstractExecutorService;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.Executor;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.RejectedExecutionException;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicIntegerFieldUpdater;
+import java.util.function.IntUnaryOperator;
+
+/**
+ * View of a cached executor service which avoids lock contention in the common path. This allows us to
+ * provide references to the same underlying pool of threads to different consumers and utilize distinct
+ * thread name patterns and instrumentation without duplicating resources. We take advantage of the fact that
+ * cached executors never queue, and always submit work immediately to avoid implementing queueing.
+ *
+ * @author <a href="mailto:ckozak@ckozak.net">Carter Kozak</a>
+ */
+final class CachedExecutorView extends AbstractExecutorService {
+    private static final AtomicIntegerFieldUpdater<CachedExecutorView> stateUpdater =
+            AtomicIntegerFieldUpdater.newUpdater(CachedExecutorView.class, "state");
```

Comment:
I'll plan to replace this with unsafe for a bit of additional optimization and uniformity with the rest of the project.